eyeball-im/src/vector/entry.rs

```diff
@@ -61,6 +54,95 @@ impl<T> Deref for ObservableVectorEntry<'_, T> {
     type Target = T;
 
     fn deref(&self) -> &Self::Target {
-        &self.inner[self.index]
+        &self.inner[self.index.value()]
+    }
+}
+
+struct EntryIndex<'a>(EntryIndexInner<'a>);
+
+enum EntryIndexInner<'a> {
+    Borrowed(&'a mut usize),
+    Owned(usize),
+}
+
+impl<'a> EntryIndex<'a> {
+    fn borrowed(index: &'a mut usize) -> Self {
+        Self(EntryIndexInner::Borrowed(index))
+    }
+
+    fn owned(index: usize) -> Self {
+        Self(EntryIndexInner::Owned(index))
+    }
+
+    fn value(&self) -> usize {
+        match &self.0 {
+            EntryIndexInner::Borrowed(idx) => **idx,
+            EntryIndexInner::Owned(idx) => *idx,
+        }
+    }
+
+    /// Remove the association with the externally-stored index, if any.
+    ///
+    /// Returns the index value for convenience.
+    fn make_owned(&mut self) -> usize {
+        match &mut self.0 {
+            EntryIndexInner::Borrowed(idx) => {
+                let idx = **idx;
+                // This line is why there's a separate EntryIndexInner type.
+                // If EntryIndex itself was the enum, this line would run Drop
+                // on the previous self in Borrowed state and bump the index,
+                // which is exactly what this method is meant to avoid.
+                self.0 = EntryIndexInner::Owned(idx);
+                idx
+            }
+            EntryIndexInner::Owned(idx) => *idx,
+        }
+    }
+}
+
+impl<'a> Drop for EntryIndex<'a> {
+    fn drop(&mut self) {
+        // If there is an association with an externally-stored index, that
+        // index must be incremented on drop. This allows an external iterator
+        // that produces ObservableVectorEntry items to advance conditionally.
+        //
+        // There are two cases this branch is not hit:
+        //
+        // - make_owned was previously called (used for removing the item and resuming
+        //   iteration with the same index)
+        // - the ObservableVectorEntry was created with ObservableVector::entry, i.e.
+        //   it's not used for iteration at all
+        if let EntryIndexInner::Borrowed(idx) = &mut self.0 {
+            **idx += 1;
```

Comment:
I would expect this to be the responsibility of the thing returned by `ObservableVectorEntries::next()`, so `ObservableVectorEntry`. It's mostly a matter of shuffling where the code is, but it seems more obvious to me to have it in the "shallower" type (`ObservableVectorEntry`) rather than the deeper type (`EntryIndex`).

Reply:
Oh right, then we don't need the extra nesting! I'll change this.